xmodule/services.py

```diff
@@ -308,3 +312,152 @@ def _handle_deprecated_progress_event(self, block, event):
         # in order to avoid duplicate work and possibly conflicting semantics.
         if not getattr(block, 'has_custom_completion', False):
             self.completion_service.submit_completion(block.scope_ids.usage_id, 1.0)
+
+
+class ProblemFeedbackService(Service):
+    """
+    An XBlock Service that allows XModules to define:
+
+    - if the answers are visible
+    - whether correctness is currently hidden. When correctness is hidden, this
+    limits the user's access to the correct/incorrect flags, messages, problem
+    scores, and aggregate subsection and course grades.
+    """
+    def __init__(self, block=None, user_is_staff=False, **kwargs):
+        super().__init__(**kwargs)
+        self._user_is_staff = user_is_staff
+        # This is needed because the `Service` class initialization expects the XBlock passed as an `xblock` keyword
+        #  argument, but the `service` method from the `DescriptorSystem` passes a `block`.
+        self._xblock = self.xblock() or block
+
+    def is_past_due(self):
+        """
+        Returns if the problem is past its due date.
+        """
+        if self._xblock:
+            return (self._xblock.close_date is not None and
+                    datetime.now(UTC) > self._xblock.close_date)
+        return False
+
+    def is_attempted(self):
+        """
+        Has the problem been attempted?
+        """
+        if self._xblock:
+            return self._xblock.attempts > 0
+        return False
+
+    def is_correct(self):
+        """
+        True if full points
+        """
+        if self._xblock:
+            # self._xblock.score is initialized in self._xblock.lcp but in this
+            # method is accessed before self._xblock.lcp so just call it first.
+            self._xblock.lcp  # pylint: disable=pointless-statement
+            return self._xblock.score.raw_earned == self._xblock.score.raw_possible
+        return False
```

Comment:
This is something that has to stay in ProblemBlock because `.lcp` is specific to that block type (and other block types may calculate things differently). If there's a service method that needs this value, please have the calling XBlock pass it in.